File: app/routers/client_debts.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session, joinedload
from sqlalchemy import func
from typing import Optional
from datetime import date

from ..database import (
    get_db, User, Client, Invoice, InvoiceItem, ClientDebt
)
from ..auth import get_current_user

router = APIRouter(prefix="/api/clients", tags=["client_debts"]) 

@router.get("/{client_id}/debts")
async def get_client_debts(
    client_id: int,
    status: Optional[str] = None,
    date_from: Optional[str] = None,
    date_to: Optional[str] = None,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    cl = db.query(Client).filter(Client.client_id == client_id).first()
    if not cl:
        raise HTTPException(status_code=404, detail="Client non trouvé")

    remaining_sql = func.coalesce(Invoice.remaining_amount, Invoice.total - func.coalesce(Invoice.paid_amount, 0))
    inv_q = (
        db.query(Invoice)
        .options(joinedload(Invoice.items))
        .filter(Invoice.client_id == client_id)
        .filter(remaining_sql > 0)
        .order_by(Invoice.date.desc())
    )
    if date_from:
        try:
            inv_q = inv_q.filter(Invoice.date >= date_from)
        except Exception:
            pass
    if date_to:
        try:
            inv_q = inv_q.filter(Invoice.date <= date_to)
        except Exception:
            pass

    invoices = []
    today = date.today()
    for inv in inv_q.all():
        amount = float(inv.total or 0)
        paid = float(inv.paid_amount or 0)
        remaining = float(inv.remaining_amount if inv.remaining_amount is not None else max(0.0, amount - paid))
        overdue = bool(inv.due_date and getattr(inv.due_date, 'date', lambda: inv.due_date)() < today and remaining > 0)
        st = "paid" if remaining <= 0 else ("overdue" if overdue else ("partial" if paid > 0 else "pending"))
        if status and st != status:
            continue
        items = [
            {
                "item_id": it.item_id,
                "product_id": it.product_id,
                "product_name": it.product_name,
                "quantity": int(it.quantity or 0),
                "price": float(it.price or 0),
                "total": float(it.total or 0),
            }
            for it in (inv.items or [])
        ]
        invoices.append({
            "id": int(inv.invoice_id),
            "invoice_number": inv.invoice_number,
            "date": inv.date,
            "due_date": inv.due_date,
            "amount": amount,
            "paid_amount": paid,
            "remaining_amount": remaining,
            "status": st,
            "items": items,
        })

    remaining_cd = func.coalesce(ClientDebt.remaining_amount, ClientDebt.amount - func.coalesce(ClientDebt.paid_amount, 0))
    cd_q = (
        db.query(ClientDebt)
        .filter(ClientDebt.client_id == client_id)
        .filter(remaining_cd > 0)
        .order_by(ClientDebt.date.desc())
    )
    if date_from:
        try:
            cd_q = cd_q.filter(ClientDebt.date >= date_from)
        except Exception:
            pass
    if date_to:
        try:
            cd_q = cd_q.filter(ClientDebt.date <= date_to)
        except Exception:
            pass

    manual_debts = []
    for d in cd_q.all():
        amount = float(d.amount or 0)
        paid = float(d.paid_amount or 0)
        remaining = float(d.remaining_amount if d.remaining_amount is not None else amount - paid)
        overdue = bool(d.due_date and getattr(d.due_date, 'date', lambda: d.due_date)() < today and remaining > 0)
        st = d.status or ("paid" if remaining <= 0 else ("overdue" if overdue else ("partial" if paid > 0 else "pending")))
        if status and st != status:
            continue
        manual_debts.append({
            "id": int(d.debt_id),
            "reference": d.reference,
            "date": d.date,
            "due_date": d.due_date,
            "amount": amount,
            "paid_amount": paid,
            "remaining_amount": remaining,
            "status": st,
            "description": d.description,
        })

    total_amount = sum(x.get("amount", 0.0) for x in invoices) + sum(x.get("amount", 0.0) for x in manual_debts)
    total_paid = sum(x.get("paid_amount", 0.0) for x in invoices) + sum(x.get("paid_amount", 0.0) for x in manual_debts)
    total_remaining = sum(x.get("remaining_amount", 0.0) for x in invoices) + sum(x.get("remaining_amount", 0.0) for x in manual_debts)

    overdue_count = sum(1 for x in invoices if x.get("status") == "overdue") + sum(1 for x in manual_debts if x.get("status") == "overdue")

    return {
        "client": {
            "client_id": cl.client_id,
            "name": cl.name,
            "email": cl.email,
            "phone": cl.phone,
        },
        "summary": {
            "total_amount": total_amount,
            "total_paid": total_paid,
            "total_remaining": total_remaining,
            "overdue_count": overdue_count,
        },
        "invoices": invoices,
        "manual_debts": manual_debts,
    }

File: static/js/client_debts.js
(function () {
  function qp(name) { try { return new URLSearchParams(window.location.search).get(name); } catch (e) { return null; } }
  function formatCurrency(v) { try { return new Intl.NumberFormat('fr-FR', { style: 'currency', currency: 'XOF', maximumFractionDigits: 0 }).format(v || 0); } catch (e) { return (v || 0) + ' XOF'; } }
  function formatDate(s) { if (!s) return '-'; try { const d = new Date(s); if (isNaN(d)) return '-'; return d.toLocaleDateString('fr-FR', { year: 'numeric', month: 'short', day: 'numeric' }); } catch (e) { return '-'; } }
  function escapeHtml(s) { return (s || '').replace(/[&<>"']/g, c => ({ "&": "&amp;", "<": "&lt;", ">": "&gt;", "\"": "&quot;", "'": "&#39;" }[c])); }

  async function loadData() {
    const cid = qp('client_id');
    if (!cid) { renderError('Client manquant.'); return; }
    try {
      const { data } = await axios.get(`/api/clients/${encodeURIComponent(cid)}/debts/`);
      renderAll(data);
    } catch (e) {
      renderError(e?.response?.data?.detail || 'Erreur lors du chargement');
    }
  }

  function renderError(msg) {
    const root = document.getElementById('clientDebtsRoot');
    if (!root) return; root.innerHTML = `<div class="alert alert-danger">${escapeHtml(String(msg || ''))}</div>`;
  }

  function renderAll(payload) {
    if (!payload) return renderError('Données indisponibles');
    const { client, summary, invoices, manual_debts } = payload;
    const invoicesFiltered = (invoices || []).filter(x => (x && x.status) !== 'paid');
    const manualDebtsFiltered = (manual_debts || []).filter(x => (x && x.status) !== 'paid');

    const hdr = document.getElementById('clientHeader');
    if (hdr) {
      hdr.innerHTML = `
        <div class="d-flex flex-wrap justify-content-between align-items-center">
          <div>
            <h2 class="h4 mb-1"><i class="bi bi-person me-2"></i>${escapeHtml(client?.name || 'Client')}</h2>
            <div class="text-muted small">${escapeHtml(client?.phone || '')} ${client?.email ? ' • ' + escapeHtml(client.email) : ''}</div>
          </div>
          <div class="d-flex gap-2">
            <a class="btn btn-outline-secondary" href="${appPath('/clients/detail?id=' + encodeURIComponent(client.client_id))}"><i class="bi bi-box-arrow-up-right me-1"></i>Fiche client</a>
            <button class="btn btn-primary" id="printBtn"><i class="bi bi-printer me-1"></i>Imprimer le récapitulatif</button>
          </div>
        </div>`;
    }

    const cards = document.getElementById('summaryCards');
    if (cards) {
      cards.innerHTML = `
        <div class="col-md-3 mb-3"><div class="card text-white debt-client"><div class="card-body"><div class="d-flex justify-content-between"><div><h4 class="mb-0">${formatCurrency(summary?.total_amount || 0)}</h4><p class="mb-0">Montant total</p></div><i class="bi bi-cash-coin display-6"></i></div></div></div></div>
        <div class="col-md-3 mb-3"><div class="card text-white debt-paid"><div class="card-body"><div class="d-flex justify-content-between"><div><h4 class="mb-0">${formatCurrency(summary?.total_paid || 0)}</h4><p class="mb-0">Total payé</p></div><i class="bi bi-check2-circle display-6"></i></div></div></div></div>
        <div class="col-md-3 mb-3"><div class="card text-white debt-remaining"><div class="card-body"><div class="d-flex justify-content-between"><div><h4 class="mb-0">${formatCurrency(summary?.total_remaining || 0)}</h4><p class="mb-0">Solde restant</p></div><i class="bi bi-exclamation-triangle display-6"></i></div></div></div></div>
        <div class="col-md-3 mb-3"><div class="card text-white debt-supplier"><div class="card-body"><div class="d-flex justify-content-between"><div><h4 class="mb-0">${parseInt(summary?.overdue_count || 0)}</h4><p class="mb-0">Échéances dépassées</p></div><i class="bi bi-alarm display-6"></i></div></div></div></div>`;
    }

    const invTbody = document.getElementById('invoicesTbody');
    if (invTbody) {
      const rows = (invoicesFiltered).map(inv => {
        const items = (inv.items || []).map(it => `${escapeHtml(it.product_name)} × ${it.quantity} — ${formatCurrency(it.total)}`).join('<br>');
        const statusBadge = badgeFor(inv.status);
        return `<tr>
          <td><strong>${escapeHtml(inv.invoice_number || String(inv.id))}</strong></td>
          <td>${formatDate(inv.date)}</td>
          <td>${inv.due_date ? formatDate(inv.due_date) : '-'}</td>
          <td>${formatCurrency(inv.amount)}</td>
          <td class="text-success">${formatCurrency(inv.paid_amount || 0)}</td>
          <td class="text-${(inv.remaining_amount || 0) > 0 ? 'danger' : 'success'}">${formatCurrency(inv.remaining_amount || 0)}</td>
          <td>${statusBadge}</td>
          <td class="small">${items || '-'}</td>
        </tr>`;
      }).join('');
      invTbody.innerHTML = rows || `<tr><td colspan="8" class="text-center text-muted py-3">Aucune facture en attente</td></tr>`;
    }

    const mdTbody = document.getElementById('manualDebtsTbody');
    if (mdTbody) {
      const rows = (manualDebtsFiltered).map(d => {
        const statusBadge = badgeFor(d.status);
        return `<tr>
          <td><strong>${escapeHtml(d.reference || String(d.id))}</strong></td>
          <td>${formatDate(d.date)}</td>
          <td>${d.due_date ? formatDate(d.due_date) : '-'}</td>
          <td>${formatCurrency(d.amount)}</td>
          <td class="text-success">${formatCurrency(d.paid_amount || 0)}</td>
          <td class="text-${(d.remaining_amount || 0) > 0 ? 'danger' : 'success'}">${formatCurrency(d.remaining_amount || 0)}</td>
          <td>${statusBadge}</td>
          <td class="small">${escapeHtml(d.description || '-')}</td>
        </tr>`;
      }).join('');
      mdTbody.innerHTML = rows || `<tr><td colspan="8" class="text-center text-muted py-3">Aucune créance manuelle en attente</td></tr>`;
    }

    const printBtn = document.getElementById('printBtn');
    if (printBtn) {
      printBtn.addEventListener('click', function () {
        const src = `/clients/debts/print/${encodeURIComponent(client.client_id)}`;
        const iframe = document.createElement('iframe');
        iframe.style.position = 'fixed';
        iframe.style.right = '0';
        iframe.style.bottom = '0';
        iframe.style.width = '0';
        iframe.style.height = '0';
        iframe.style.border = '0';
        iframe.src = src;
        iframe.onload = function () {
          try { iframe.contentWindow?.focus(); iframe.contentWindow?.print(); } catch (e) { }
          setTimeout(() => { try { iframe.remove(); } catch (_) { } }, 2000);
        };
        document.body.appendChild(iframe);
      });
    }
  }

  function badgeFor(st) {
    const map = { paid: 'bg-success', partial: 'bg-warning text-dark', overdue: 'bg-danger', pending: 'bg-secondary' };
    const label = { paid: 'Payé', partial: 'Partiel', overdue: 'En retard', pending: 'En attente' };
    const cls = map[st] || 'bg-secondary';
    const txt = label[st] || st;
    return `<span class="badge ${cls}">${txt}</span>`;
  }

  document.addEventListener('DOMContentLoaded', loadData);
})();
